hdKeysGenerator/main.go

```diff
@@ -0,0 +1,180 @@
+	if err != nil {
+		log.Error(err.Error())
+		os.Exit(1)
+	}
+
+	log.Info("Done in", "seconds", time.Since(start).Seconds())
+}
+
+func generateKeys(ctx *cli.Context) error {
+	cliFlags := getParsedCliFlags(ctx)
+
+	constraints, err := newConstraints(cliFlags.numShards, cliFlags.actualShard, cliFlags.projectedShard)
+	if err != nil {
+		return err
+	}
+
+	exporter, err := export.NewExporter(export.ArgsNewExporter{
+		ActualShard:    cliFlags.actualShard,
+		ProjectedShard: cliFlags.projectedShard,
+		StartIndex:     cliFlags.startIndex,
+		NumKeys:        int(cliFlags.numKeys),
+		Format:         cliFlags.exportFormat,
+		OutputFile:     cliFlags.outputFile,
+	})
+	if err != nil {
+		return err
+	}
+
+	mnemonic, err := askMnemonic()
+	if err != nil {
+		return err
+	}
+
+	args := argsGenerateKeysInParallel{
+		numTasks:        cliFlags.numTasks,
+		startIndex:      cliFlags.startIndex,
+		useAccountIndex: cliFlags.useAccountIndex,
+		numKeys:         int(cliFlags.numKeys),
+		constraints:     constraints,
+	}
+
+	generatedKeys, err := generateKeysInParallel(context.Background(), args, mnemonic)
+	if err != nil {
+		return err
+	}
+
+	log.Info("done key generation")
+
+	err = exporter.ExportKeys(generatedKeys)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func generateKeysInParallel(
+	ctx context.Context,
+	args argsGenerateKeysInParallel,
+	mnemonic data.Mnemonic,
+) ([]common.GeneratedKey, error) {
+	allGeneratedKeys := make([]common.GeneratedKey, 0, args.numKeys)
+
+	numKeys := int(args.numKeys)
+	numTasks := args.numTasks
+	slidingIndex := args.startIndex
+
+	// Description of the parallel model:
+	// In a loop, as long as there are keys to be generated:
+	//	- a number of "numTasks" (parallel) tasks are created.
+	// 	- each task checks a number of "fixedTaskSize" indexes account / address indexes for eligibility.
+	//  - the output (selected indexes, generated keys) is accumulated in a slice
+	// At the end of the loop, the accumulator slice is sorted by account & address indexes,
+	// while redundantly-generated keys are ignored.
+	//
+	// This parallelization model leads to some redundant work when a small number of keys are requested
+	// and / or the generation constraints are "easy" (they do not imply a lot of index skipping).
+	// However, the model should behave well when a lot of account / address indexes have to be checked for eligibility.
+	for len(allGeneratedKeys) < numKeys {
+		generatedKeysByTask := make([][]common.GeneratedKey, numTasks)
+		tasks, newSlidingIndex := createTasks(argsCreateTasks{
+			numTasks:        numTasks,
+			startIndex:      slidingIndex,
+			useAccountIndex: args.useAccountIndex,
+			constraints:     args.constraints,
+		})
+
+		errs, _ := errgroup.WithContext(ctx)
+
+		for taskIndex, task := range tasks {
+			i := taskIndex
+			t := task
+
+			errs.Go(func() error {
+				keys, err := t.doGenerateKeys(mnemonic)
+				if err != nil {
+					return err
+				}
+
+				generatedKeysByTask[i] = keys
```

Comment:
I have a concern of race condition here. We can simplify this to work with channels. Multiple production go routines and one go routine that collects the results.